src/hyperion/device_setup_plans/check_topup.py

```diff
@@ -0,0 +1,80 @@
+import bluesky.plan_stubs as bps
+from dodal.devices.detector import DetectorParams
+from dodal.devices.synchrotron import Synchrotron, SynchrotronMode
+
+from hyperion.log import LOGGER
+
+ALLOWED_MODES = [SynchrotronMode.USER.value, SynchrotronMode.SPECIAL.value]
+DECAY_MODE_CTDW = -1
+
+
+def _in_decay_mode(time_to_topup):
+    if time_to_topup == DECAY_MODE_CTDW:
+        LOGGER.info("Machine in decay mode, gating disabled")
+        return True
+    return False
+
+
+def _gating_permitted(machine_mode):
+    if machine_mode in ALLOWED_MODES:
+        LOGGER.info("Machine in allowed mode, gating top up enabled.")
+        return True
+    LOGGER.info("Machine not in allowed mode, gating disabled")
+    return False
+
+
+def _delay_to_avoid_topup(total_run_time, time_to_topup):
+    if total_run_time > time_to_topup:
+        LOGGER.info(
+            """
+            Total run time for this collection exceeds time to next top up.
+            Collection delayed until top up done.
+            """
+        )
+        return True
+    LOGGER.info(
+        """
+        Total run time less than time to next topup. Proceeding with collection.
+        """
+    )
+    return False
+
+
+def wait_for_topup_complete(synchrotron):
+    start = yield from bps.rd(synchrotron.top_up.start_countdown)
+    while start == 0:
+        yield from bps.sleep(0.1)
+        start = yield from bps.rd(synchrotron.top_up.start_countdown)
+
+
+def check_topup_and_wait_if_necessary(
+    synchrotron: Synchrotron,
+    params: DetectorParams,
```

Comment:
Should: It's possible that the detector parameters doesn't actually have the correct number of frames in when we're free running the detetcor. I think it would be better for this function just to take the expected time the collection will take, the individual collections can calculate it